### PR TITLE
add optional logging

### PR DESCRIPTION
**Pull request by carns**
In GitLab by @roblatham00 on Sep 23, 2020, 18:06

I added some additional logging to abt-io routines -- I wanted to verify my code was reading/writing what I thought it was. Maybe non-mpi darshan is a better way to go...

### CODE REVIEW

**Comment by carns** [comment]
In GitLab by @carns on Sep 24, 2020, 08:32

Looks great to me.  I like the idea of having something simple like this you can just turn on without doing anything extra.

My only minor suggestion would be rather than check getenv() in abt_io_log(), check it once in init and then set a flag in the instance (and check the flag in abt_io_log).  I don't know if there is any meaningful cost associated with getenv() calls but that would cut them back, and having this as a flag in the abtio instance will make it straightforward to integrate into the json config when it lands.

**Comment by carns** [comment]
In GitLab by @roblatham00 on Sep 24, 2020, 16:32

added 1 commit

<ul><li>fbda96a1 - add optional logging</li></ul>

[Compare with previous version](/sds/abt-io/-/merge_requests/3/diffs?diff_id=10755&start_sha=376430df6b50106f6006c05d2480bb7744052ae6)

**Comment by carns** [comment]
In GitLab by @roblatham00 on Sep 24, 2020, 16:38

Latest revision ended up being a bit more invasive. 

Putting a check in `abt_io_init()` let me turn the times into "times since init" not "time since unix epoch" which is more like what darshan does and more human readable.

**Comment by carns** [comment]
In GitLab by @roblatham00 on Sep 28, 2020, 17:57

Hm.  I wish I had some good way to tell abt-io which rank it was.  Right now, output from N processes will get interleaved.  maybe the environment variable should take a file name specifier e.g. ABT_IO_TRACE=server.%p.out (or '-' for stderr) and log to that.

**Comment by carns** [comment]
In GitLab by @roblatham00 on Sep 29, 2020, 11:30

added 6 commits

<ul><li>fbda96a1...85807d78 - 5 commits from branch <code>master</code></li><li>2e2dfb7e - add optional logging</li></ul>

[Compare with previous version](/sds/abt-io/-/merge_requests/3/diffs?diff_id=10774&start_sha=fbda96a1e5d7a660796868aa674cf16e53c70da2)

**Comment by carns** [comment]
In GitLab by @roblatham00 on Sep 29, 2020, 15:19

Updated for json: no more environment variable

**Comment by carns** [comment]
In GitLab by @carns on Nov 26, 2020, 10:56

Sorry @roblatham00 , breaking this MR once again if !4 goes to master (converting to json-c and a json format that fits better with bedrock).  I can help update if/when that one lands.

Also noting that @mdorier has added logging to Margo that looks like this: https://xgitlab.cels.anl.gov/sds/abt-io/-/merge_requests/4  I don't think the abt-io logging needs to follow the same conventions (the abt-io logging is really more like I/O tracing), just pointing it out in case there is anything helpful there.

**Comment by carns** [comment]
In GitLab by @mdorier on Nov 26, 2020, 11:00

I didn't see this PR until now but if ABT-IO could have logging by dependency injection (like Margo) that would be useful: Bedrock is injecting its own logging functions into Margo to unify their output into the same logger.

**Comment by roblatham00** [comment]
Given a json config like this:
```
% cat abt.json
{
  "version":"0.5.1",
  "trace_io":true,
  "null_io_write":false,
  "null_io_read":false,
  "internal_pool_flag":true,
  "internal_pool":{
    "num_xstreams":16,
    "kind":"fifo_wait",
    "access":"mpmc"
  }
}
```
output looks like this:
```
% ./tests/basic
% ./tests/basic abt.json
#Module Rank    Op      Segment Offset  Length  Start(s)        End(s)
X_ABTIO -1      mkostemp        -1      0       0       0.019910        0.019987
X_ABTIO -1      pwrite  -1      0       4       0.020215        0.020270
X_ABTIO -1      fallocate       -1      4       4       0.020380        0.022147
X_ABTIO -1      truncate        -1      0       1024    0.022354        0.022393
X_ABTIO -1      close   -1      0       0       0.022542        0.022584
X_ABTIO -1      stat    -1      0       0       0.022752        0.022782
X_ABTIO -1      statfs  -1      0       0       0.022978        0.023006
X_ABTIO -1      unlink  -1      0       0       0.023125        0.023235
```

**Comment by carns** [comment]
The README.md describes setting the ABT_IO_TRACE environment variable, but it doesn't look like that's implemented?

An env override for the json would probably be convenient (maybe named ABT_IO_TRACE_IO to more closely match the corresponding json parameter?  Or else take the _io suffix off of both).  There are some examples in Margo that do that (environment variables that can override and set json parameters):

https://github.com/mochi-hpc/mochi-margo/blob/main/src/margo-init.c#L488

**Comment by roblatham00** [comment]
thanks for the pointer to margo.  took the same approach in abt-io

File: src/abt-io.c [hunk 1]
```diff
@@ -1578,6 +1639,20 @@ static int validate_and_complete_config(struct json_object* _config,
     if (!CONFIG_HAS(_config, "null_io_read", val)) {
         CONFIG_OVERRIDE_BOOL(_config, "null_io_read", 0, "null_io_read", 0);
     }
+
+    {
+        const char* trace_io_str = getenv("ABT_IO_TRACE_IO");
+        if (!CONFIG_HAS(_config, "trace_io", val)) {
+            CONFIG_OVERRIDE_BOOL(_config, "trace_io", 0, "trace_io", 0);
+            if (trace_io_str)
+                CONFIG_OVERRIDE_BOOL(_config, "trace_io", trace_io_enabled,
```
Comment (carns) [inline src/abt-io.c:1648]: I'm getting a compile problem on this: 
```
../src/abt-io.c:1649:50: error: ‘trace_io_enabled’ undeclared (first use in this function)
```

**Comment by roblatham00** [comment]
crud.  deleted one too many lines when I cleaned out printfs..  should be better now plus a bonus patch that fixes a "-Wall" warning

**Comment by carns** [comment]
Great, thanks @roblatham00.

I think the option mentioned in the comment (about being able to output the trace to a file) will be handy at some point, but no need to wait for that to merge.

**Pull request merged**